### PR TITLE
Multiauth update email ui

### PR DESCRIPTION
**Pull request by maddiedierker**
Depends on #23143. 

This PR updates the front-end to honor the new `PATCH /users/set_email` functionality that allows all migrated users (except word/picture accounts) to update their email address from the account details page.

### Details
- in `ChangeEmailModal`: For migrated users who have an `encrypted_password`, the password field is still a required input when the user is changing their email address. However, for users who can now update their email address but do not have a Code.org password (oauth-only users), the password field is no longer required.
- on "Edit Account Details" view: For migrated users, the password field will only be shown if: 1) the user already has a password, OR 2) the user is a word/picture account. In these cases, an editable input and a "See your teacher to manage your password" message are still shown, respectively. Otherwise, this field is not shown and will be replaced by the "Add a password" feature (coming next!)

### Example
For an oauth-only migrated user, the UI looks like...

"Update" link for email and hidden password field:
screen shot 2018-06-19 at 10 19 01 am

No password field in the `ChangeEmailModal`:
screen shot 2018-06-19 at 10 19 11 am

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -772,7 +772,11 @@ def self.from_omniauth(auth, params)
   end
 
   def oauth?
-    OAUTH_PROVIDERS.include? provider
+    if migrated?
+      authentication_options.any? {|auth| OAUTH_PROVIDERS.include? auth.credential_type}
+    else
+      OAUTH_PROVIDERS.include? provider
+    end
```
Comment (maddiedierker) [inline dashboard/app/models/user.rb:779]: copied from @ewjordan's PR (#23136)

**Comment by islemaster** [comment]
Design nits: I find this a little confusing, and would prefer the new note was inline with the "Email" label.  I'd also prefer we used sentence case for this note.

image

**Comment by maddiedierker** [comment]
@islemaster re: design nit -- i agree that the email hint looks a little out of place. i was trying to keep with the style of our current hints, which look like:
screen shot 2018-06-19 at 11 30 35 am
i'd like them to be consistent, but maybe it's okay if this one is different since it is already contextually different?

re: changing `ChangeEmailModal`, it was really easy! there were a few layers that the new prop needed to be passed through, but it was really easy to work with and update with this change 🎉

**Comment by islemaster** [comment]
(Talked in person) Leave the style thing for now - we'll get feedback at the bug bash Thursday and do a style pass.

Glad to hear I wrote usable code!

**Pull request merged**